### DISCUSSION TITLE
Exploring the use of this Library to connect with an esp32 via Bluetooth low energy

### DISCUSSION THREAD

**Discussion by Amritanshu-Shandilya** [kevincar/bless#112] [General]
Hi. I am trying to make a smartwatch using esp32. This is the code [repository](https://github.com/Amritanshu-Shandilya/Mini_watch)
Also to complicate things and to open an entirely new domain of functionalities I planned to write an companion app for the smart watch in python(, scary right? It is). As you might have guessed it why I wanted to use bless . I am using beeware for developing the companion app and it is still in its infancy, you could have a look of it if you want :
[APP](https://github.com/Amritanshu-Shandilya/mini-watch-companion)

I have some questions like :

have you used / or came across someone who has used bless for connecting with a microcontroller ? Using ble
does bless works for android? like is it capable of accessing Bluetooth hardware of smartphones?
also many more. I hope you would reply to this as soon as possible. It would be very helpful for my project if you could connect with me here.
Also this isnt a commercial project. Neither it is an assignment for my college. I am just building this stuff because I want to learn stuff like this and one day develop something worth. Have a nice day. Thank you.

**Comment by kevincar** [comment]
Hi @Amritanshu-Shandilya 

> I planned to write an companion app for the smart watch in python.

Are you planning to run this companion app on a desktop environment or another esp32-like microcontroller? What role are you seeking your watch and your companion app to play for Bluetooth Low energy? Would you like the watch or the companion app to act as the bluetooth peripheral (server)?

> Have you used or came across someone who has used bless for connecting with a microcontroller? 

I have not. Bless is currently implemented to run on windows, macOS, and linux. Without diving into details, I would imagine that their either needs to be some Espressif compiler for python or are you using CircuitPython to run python on the microcontroller natively? 

> Using ble does bless works for android? Like is it capable of accessing Bluetooth hardware of smartphones?

Currently bless does not provide support for mobile devices. However, [`bleak`](https://github.com/hbldh/bleak/tree/develop) does run on android if you'd like to implement a BLE client. Bless does run on CoreBluetooth, so I would image that Python-enabled systems to run as iOS apps should also work.

**Reply by Amritanshu-Shandilya** [reply]
Hi @kevincar thanks for your quick response.

`Are you planning to run this companion app on a desktop environment or another Esp32-like micro-controller? What role are you seeking your watch and your companion app to play for Bluetooth Low energy? Would you like the watch or the companion app to act as the Bluetooth peripheral (server)?`

No I was planning for an android application that talks to Esp32 micro controller through Bluetooth Low Energy. My goal was to make something like Apple Watch and iPhone : introduce a communication link between a smartwatch (that doesn't have enough resources) and a smartphone so that smartphone can share some load of the micro-controller and also provide some additional functions like navigation. And for this I wanted to setup my smartphone as a server because it has more powerful resources when compared to a simple micro-controller like ESP32.

`Bless is currently implemented to run on windows, mac OS, and Linux.`
(sad emoji)

` I would imagine that their either needs to be some Espressif compiler for python or are you using CircuitPython to run python on the micro-controller natively?`
No I am using Arduino Framework for developing the program for the ESP32. I have asked some more experienced developers about this compatibility issue between two different languages and they said that it is fine since I need to program some special signals that both programs can understand, that's why I am looking for a python library for BLE support. 

`[bleak](https://github.com/hbldh/bleak/tree/develop) does run on android if you'd like to implement a BLE client.`
Yes bleak was my first choice but I had to backup because of two reasons:

1.  It creates a client only
2.  It didn't support python's Beeware framework. (But this isnt an issue any longer as I am ready to shift to kivy)

**Reply by kevincar** [reply]
> No I was planning for an android application that talks to Esp32 micro controller through Bluetooth Low Energy

So you are using the Arduino Framework on the ESP32 (smartwatch) and KIVY for android (companion app)? If this is the case then I would suggest that the smartwatch act as the server fully implemented in Arduino C++. It seems like you can use the [ESP32 BLE Library](https://github.com/nkolban/ESP32_BLE_Arduino). The [BLEServer class](https://github.com/nkolban/ESP32_BLE_Arduino/blob/adc2aee2f0d01eb6b30dd5ad3589f2cc89934beb/src/BLEServer.h#L63) seems like a promising starting point. Then on the client side, if you're using KIVY for building a python-based android app, you can use bleak as the client to search for and interact with the peripheral device (smart watch). 

Is this not a viable option for you?

**Reply by Amritanshu-Shandilya** [reply]
Hi Kevin. Sorry I was busy in some personal tasks and therefore I could not get in touch with you earlier.

I was thinking of doing the same but there is a little bit problem in this setup as per my understanding of BLE, please correct me if I am wrong. The Smartphone is a more powerful device and has a lot more sensors processing power memory and battery than an esp32. So I wanted my phone to do heavy lifting for me like passing data like time date GPS etc to the ESP32 and thus I should make the smartphone the server in this architecture. ESP as the client would just receive this data broadcasted and use it to display some stuff on the screen like the time and Directions etc.

I am still reading through the BLE primer so I am not sure if my traditional understanding of BLE server & client is correct or not. Please do correct me if I have the wrong idea. Thank You!

**Reply by kevincar** [reply]
> The Smartphone is a more powerful device ... I wanted my phone to do heavy lifting ... thus I should make the smartphone the server

In BLE, servers are the devices that broadcast/advertise services the ability to connect (these are actually known as peripheral devices), and the client is the device that actually does the connecting (these are known as central devices). For example, when pairing a smart phone to bluetooth headphones, the headphones advertise their availability and the phone selects them for connecting. In this sense, the headphones are the peripheral device (i.e., the server) that broadcasts some service (such as playing audio) and the phone is the central device. 

Since your watch is simply displaying information sent by the phone it may make sense to have the ESP smart wash be the peripheral device (server) while the smart phone is the central (client). In the traditional sense of servers/clients my intuition would be that servers need to do heavy lifting because they service many clients, sometimes simultaneously (think web servers). In the case of BLE, even if the peripheral devices services many clients, it's not to the same degree as a web server. Hence it is a peripheral and central device rather than server and client.

**Reply by Amritanshu-Shandilya** [reply]
Thanks for helping me out here. I guess the idea suggested to you might work. The smartwatch can become a server advertising its availability and the app could be a client. I had talked with the developers of BLEak, they said that their project does not support BeeWare which I was using at that time and advised me to use Kivy. I am now working on developing the Kivy Application.
